### PR TITLE
Add hidden-by-default ruff.applyFormat command

### PR DESCRIPTION
**Pull request by konstin**
This adds a very basic format command that depends on https://github.com/charliermarsh/ruff/pull/4453.

### CODE REVIEW

**Comment by konstin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #113** <a href="https://app.graphite.dev/github/pr/charliermarsh/ruff-lsp/113" target="_blank">Graphite</a> 
    * **PR #114** <a href="https://app.graphite.dev/github/pr/charliermarsh/ruff-lsp/114" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/charliermarsh/ruff-lsp/114?utm_source=stack-comment).

**Comment by MichaReiser** [comment]
I'll have a closer look tomorrow. There's one question that came to mind:
Can you add an explanation to the summary explaining your reasoning behind adding a new format comment over using the format capability https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/?

**Comment by charliermarsh** [comment]
Is the idea that we would eventually support the conventional LSP formatting actions, and that this custom action is temporary? Or that both would exist?

File: ruff_lsp/server.py
Comment (MichaReiser) [inline ruff_lsp/server.py:63]: Is this the standard logfile name when using python or what does `pygls` stand for?

File: ruff_lsp/server.py
Comment (MichaReiser) [inline ruff_lsp/server.py:607]: A bundler could be useful here so that we could have two different builds. But I don't see how we could achieve the same with python. I guess what we really want are feature flags and your `RUFF_BETA_INTERNAL` is kind of that

File: ruff_lsp/server.py
Comment (MichaReiser) [inline ruff_lsp/server.py:610]: Nit:
```suggestion
def apply_format_impl(language_server: server.LanguageServer, arguments: tuple[TextDocument]):
```

File: ruff_lsp/server.py
Comment (MichaReiser) [inline ruff_lsp/server.py:626]: What's the meaning of the `*`?

File: ruff_lsp/server.py
Comment (MichaReiser) [inline ruff_lsp/server.py:963]: Is it intentional that you removed the `\r\n` replacement? I don't think it is needed for ruff today and could, potentially, result in incorrect offsets.

**Review by MichaReiser — APPROVED** [review]
Looks good to me, assuming that the longterm plan is to replace the custom command with the LSP format capabilities. Would you mind expanding the PR summary with what your long-term plans are (e.g. This PR adds a temporary `ruff.applyFormat` command to ease testing and fast iteration. It is gated by the `ENV_...` and my long-term plan is to replace the command with the LSP-s built in format capabilities.

File: ruff_lsp/server.py
Comment (konstin) [reply ruff_lsp/server.py:63]: `pygls` is the lsp library we're using, the code code is just the example from https://pygls.readthedocs.io/en/latest/pages/advanced_usage.html?highlight=log#logging. I'm happy to change this to anything else if you have good ideas for logging, this was a very opportunistic addition because i sometimes need a log file that's independent of VS code

Comment (konstin) [reply ruff_lsp/server.py]: Good point, renamed it `ruff-lsp.log`

File: ruff_lsp/server.py
Comment (konstin) [reply ruff_lsp/server.py]: Python has extras but they unfortunately don't persist to runtime

File: ruff_lsp/server.py
Comment (konstin) [reply ruff_lsp/server.py:646]: Afterwards keyword argument only, e.g. `_fix_helper(document, only="F")` is valid but `_fix_helper(document, "F")` is not. This part i haven't change, no specific opinion given that those are internal APIs and we can change them at will.

https://stackoverflow.com/questions/59661042/what-do-single-star-and-slash-do-as-independent-parameters

File: ruff_lsp/server.py
Comment (konstin) [reply ruff_lsp/server.py:963]: I asked charlie and he said it's likely historical. I removed it because i would like avoid doing any content transformations on the LSP side, those should all be done in ruff, and i'm a bit worried differences between the file in VS Code/on disk and what ruff thinks the file can get really confusing. For the offsets we communicate line and character in the LSP protocol, so i'd hope that they are unaffected

**Comment by konstin** [comment]
Missing steps:
 * Merge https://github.com/charliermarsh/ruff/pull/4453
 * revert CI changes

**Comment by konstin** [comment]
waiting for the next ruff release to unblock CI

**Pull request merged**
